src/coreclr/jit/forwardsub.cpp

```diff
@@ -247,15 +247,23 @@ class ForwardSubVisitor final : public GenTreeVisitor<ForwardSubVisitor>
 
         if (node->OperIsLocal())
         {
-            unsigned const lclNum = node->AsLclVarCommon()->GetLclNum();
-
-            // Uses of address-exposed locals are modelled as global refs.
-            //
+            unsigned const   lclNum = node->AsLclVarCommon()->GetLclNum();
             LclVarDsc* const varDsc = m_compiler->lvaGetDesc(lclNum);
 
             if (varDsc->IsAddressExposed())
             {
-                m_accumulatedFlags |= GTF_GLOB_REF;
+                if (!user->OperIs(GT_ASG) || (node != user->gtGetOp1()))
+                {
+                    // Uses of address-exposed locals are modelled as global refs.
+                    m_accumulatedFlags |= GTF_GLOB_REF;
+                }
+                else
+                {
+                    // Stores to address-exposed locals are global refs as well,
+                    // but defer adding into the flags just yet: the actual side
+                    // effect will happen at the point of the ASG, not the LHS.
+                    user->gtFlags |= GTF_GLOB_REF;
```

Comment:
I've held back from "correcting" the tree flags in the past. I'm not fundamentally against the idea, but forward sub might now modify IR without actually doing any substitution.

So, I'd also track if any tree flags have changed and use that to set the modified phase status.

Reply:
Doesn't this flag effect updating already happen, in a more prominent manner, with the `gtUpdateStmtSideEffects(nextStmt/stmt)` calls?

That said, this parent updating is a little bit of a hack, let me see if it can be achieved in some other way.

Reply:
> Doesn't this flag effect updating already happen, in a more prominent manner, with the `gtUpdateStmtSideEffects(nextStmt/stmt)` calls?

Good point.  I recalled doing that for when substituting but not upstream where we still might bail out.